src/OVAL/oval_session.c

```diff
@@ -0,0 +1,657 @@
+{
+	__attribute__nonnull__(session);
+
+	oscap_free(session->export.syschars);
+	session->export.syschars = oscap_strdup(filename);
+}
+
+void oval_session_set_report_export(struct oval_session *session, const char *filename)
+{
+	__attribute__nonnull__(session);
+
+	oscap_free(session->export.report);
+	session->export.report = oscap_strdup(filename);
+}
+
+static bool oval_session_validate(struct oscap_source *source, xml_reporter fn, oscap_document_type_t type)
+{
+	if (oscap_source_get_scap_type(source) == type) {
+		if (oscap_source_validate(source, fn, NULL))
+			return false;
+	}
+	else {
+		oscap_seterr(OSCAP_EFAMILY_OVAL, "Type mismatch: %s. Expecting %s "
+			"but found %s.\n", oscap_source_readable_origin(source),
+			oscap_document_type_to_string(type),
+			oscap_document_type_to_string(oscap_source_get_scap_type(source)));
+		return false;
+	}
+
+	return true;
+}
+
+static int oval_session_load_definitions(struct oval_session *session, xml_reporter fn)
+{
+	__attribute__nonnull__(session);
+	__attribute__nonnull__(session->source);
+
+	struct ds_sds_session *sds_session;
+
+	if (!oval_session_validate(session->source, fn, OSCAP_DOCUMENT_SDS)) {
+		/* the source can still be OVAL Definitions */
+		oscap_clearerr();
+		if (!oval_session_validate(session->source, fn, OSCAP_DOCUMENT_OVAL_DEFINITIONS))
+			return 1;
+	}
+
+	if (oscap_source_get_scap_type(session->source) == OSCAP_DOCUMENT_SDS) {
+		if ((sds_session = ds_sds_session_new_from_source(session->source)) == NULL) {
+			return 1;
+		}
+
+		ds_sds_session_set_datastream_id(sds_session, session->datastream_id);
+		if (ds_sds_session_register_component_with_dependencies(sds_session,
+					"checks", session->component_id, "oval.xml") != 0) {
+			ds_sds_session_free(sds_session);
+			return 1;
+		}
+
+		session->oval.definitions = ds_sds_session_get_component_by_href(sds_session, "oval.xml");
+		if (session->oval.definitions == NULL) {
+			oscap_seterr(OSCAP_EFAMILY_OVAL, "Internal error: OVAL file was not found in "
+					"Source DataStream session cache!");
+			ds_sds_session_free(sds_session);
+			return 1;
+		}
+	}
+	else {
+		oscap_source_free(session->oval.definitions);
+		session->oval.definitions = session->source;
+		session->source = NULL;		/* don't free the same thing twice */
+	}
+
+	/* import OVAL Definitions */
+	if (session->def_model) oval_definition_model_free(session->def_model);
+	session->def_model = oval_definition_model_import_source(session->oval.definitions);
+	if (session->def_model == NULL) {
+		oscap_seterr(OSCAP_EFAMILY_OVAL, "Failed to import the OVAL Definitions from '%s'.\n",
+				oscap_source_readable_origin(session->oval.definitions));
+				ds_sds_session_free(sds_session);
+		return 1;
+	}
+
+	return 0;
+}
+
+static int oval_session_load_variables(struct oval_session *session, xml_reporter fn)
+{
+	__attribute__nonnull__(session);
+
+	if (session->oval.variables == NULL) {
+		oscap_seterr(OSCAP_EFAMILY_OVAL, "No OVAL Variables set.\n");
+		return 1;
```

Comment:
This is basically unreachable code.

In `oval_session_load` you only call `oval_session_load_variables` when the oval.variables are *not* NULL. What about calling `_load_variables` any time and doing no-op when os->oval.variables is NULL.

The result will be a less code to maintain.